src/Cli/dotnet/commands/Run/VirtualProjectBuildingCommand.cs

```diff
@@ -139,86 +178,443 @@ private ProjectInstance CreateProjectInstance(
-              <!-- We need to explicitly import Sdk props/targets so we can override the targets below. -->
-              <Import Project="Sdk.props" Sdk="Microsoft.NET.Sdk" />
+        var builder = ImmutableArray.CreateBuilder<CSharpDirective>();
 
-              <PropertyGroup>
-            {CommonProjectProperties}
+        // NOTE: When Roslyn is updated to support "ignored directives", we should use its SyntaxTokenParser instead.
+        foreach (var line in sourceFile.Text.Lines)
+        {
+            var lineText = sourceFile.Text.ToString(line.Span);
 
-                <EnableDefaultItems>false</EnableDefaultItems>
-              </PropertyGroup>
+            if (Patterns.Shebang.IsMatch(lineText))
+            {
+                builder.Add(new CSharpDirective.Shebang { Span = line.SpanIncludingLineBreak });
+            }
+            else if (Patterns.Directive.Match(lineText) is { Success: true } match)
+            {
+                builder.Add(CSharpDirective.Parse(sourceFile, line.SpanIncludingLineBreak, match.Groups[1].Value, match.Groups[2].Value));
+            }
+        }
 
-              <Import Project="Sdk.targets" Sdk="Microsoft.NET.Sdk" />
-
-              <!--
-                Override targets which don't work with project files that are not present on disk.
-                See https://github.com/NuGet/Home/issues/14148.
-              -->
-
-              <Target Name="_FilterRestoreGraphProjectInputItems"
-                      DependsOnTargets="_LoadRestoreGraphEntryPoints"
-                      Returns="@(FilteredRestoreGraphProjectInputItems)">
-                <ItemGroup>
-                  <FilteredRestoreGraphProjectInputItems Include="@(RestoreGraphProjectInputItems)" />
-                </ItemGroup>
-              </Target>
-
-              <Target Name="_GetAllRestoreProjectPathItems"
-                      DependsOnTargets="_FilterRestoreGraphProjectInputItems"
-                      Returns="@(_RestoreProjectPathItems)">
-                <ItemGroup>
-                  <_RestoreProjectPathItems Include="@(FilteredRestoreGraphProjectInputItems)" />
-                </ItemGroup>
-              </Target>
-
-              <Target Name="_GenerateRestoreGraph"
-                      DependsOnTargets="_FilterRestoreGraphProjectInputItems;_GetAllRestoreProjectPathItems;_GenerateRestoreGraphProjectEntry;_GenerateProjectRestoreGraph"
-                      Returns="@(_RestoreGraphEntry)">
-                <!-- Output from dependency _GenerateRestoreGraphProjectEntry and _GenerateProjectRestoreGraph -->
-              </Target>
-            </Project>
-            """;
-        ProjectRootElement projectRoot;
-        using (var xmlReader = XmlReader.Create(new StringReader(projectFileText)))
+        // The result should be ordered by source location, RemoveDirectivesFromFile depends on that.
+        return builder.ToImmutable();
+    }
+
+    public static SourceFile CreateSourceFile(string filePath)
```

Comment:
It may be good to use a verb like "Load" here as "Create" has a connotation of writing files to disk